[SPARK-26339][SQL]Throws better exception when reading files that start with underscore

KeiichiHirobe opened the pull request:
## What changes were proposed in this pull request?
As the description in [SPARK-26339](https://issues.apache.org/jira/browse/SPARK-26339),   spark.read behavior is very confusing when reading files that start with underscore,  fix this by throwing exception which message is "Path does not exist".

## How was this patch tested?
manual tests.
Both of codes below throws exception which message is "Path does not exist".
```
spark.read.csv("/home/forcia/work/spark/_test.csv")
spark.read.schema("test STRING, number INT").csv("/home/forcia/work/spark/_test.csv")
```

Commit:
Throws better exception when reading files that start with underscore

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:
```diff
@@ -554,7 +554,8 @@ case class DataSource(
       // Don't need to check once again if files exist in streaming mode
-      if (checkFilesExist && !fs.exists(globPath.head)) {
+      if (checkFilesExist &&
+          (!fs.exists(globPath.head) || InMemoryFileIndex.shouldFilterOut(globPath.head.getName))) {
```
I'm probably misunderstanding, but doesn't this still cause it to throw a 'Path does not exist' exception?

KeiichiHirobe replied:
`InMemoryFileIndex.shouldFilterOut` returns true if argument starts with underscore, so throw a 'Path does not exist' exception. I've checked and exception below was thrown.

```
org.apache.spark.sql.AnalysisException: Path does not exist: file:_test.csv;
  at org.apache.spark.sql.execution.datasources.DataSource.$anonfun$checkAndGlobPathIfNecessary$1(DataSource.scala:558)
  at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:244)
  at scala.collection.immutable.List.foreach(List.scala:392)
  at scala.collection.TraversableLike.flatMap(TraversableLike.scala:244)
  at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:241)
  at scala.collection.immutable.List.flatMap(List.scala:355)
  at org.apache.spark.sql.execution.datasources.DataSource.checkAndGlobPathIfNecessary(DataSource.scala:545)
  at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:359)
  at org.apache.spark.sql.DataFrameReader.loadV1Source(DataFrameReader.scala:231)
  at org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:219)
  at org.apache.spark.sql.DataFrameReader.csv(DataFrameReader.scala:625)
  at org.apache.spark.sql.DataFrameReader.csv(DataFrameReader.scala:478)
```

srowen replied:
I see, I didn't read carefully. This is the new desired behavior. I agree it would be better to not end up with an odd CSV parsing message. I wonder if we can clarify the message further with a different exception for the new case. The path does exist; it's just ignored.

```
if (checkFilesExist) {
  val firstPath = globPath.head
  if  (!fs.exists(firstPath)) {
    // ... Path does not exist
  } else if (shouldFilterOut...) {
    // ... Path exists but is ignored
  }
}
```

KeiichiHirobe replied:
@srowen I pushed now.  Could you please check my commit?

Commit:
Clarify the message further with a different exception for file which is ignored

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:
```diff
@@ -554,8 +554,13 @@ case class DataSource(
+        if (!fs.exists(firstPath)) {
+          throw new AnalysisException(s"Path does not exist: ${firstPath}")
+        } else if (InMemoryFileIndex.shouldFilterOut(firstPath.getName)) {
+          throw new AnalysisException(s"Path exists but is ignored: ${firstPath}")
```
Also, looks it's going to not check children.

HyukjinKwon commented:
I think one thing we could consider is to leave a trace or debug log to show which files are ignored when the file is listed.

srowen reviewed (commented):
Yeah, we may need a different approach here. Really we want to check all the files in the glob, and maybe log a debug statement about ones that are ignored, and throw an exception only if all of the files are filtered out. Otherwise this fails if the first file happens to have an underscore, but others don't, and there is a convention that `_files` are just ignored in Hadoop/Spark/etc.

That would cover the original case where the user specified one file that is ignored.

I don't even know if it's a behavior change in many cases, because several things will fail later anyway (like CSV schema inference here) if there are no files. If the user specifies a path that has only underscore files, I could see the argument that it should just produce an empty dataframe, but, that could be surprising as well if there are data files there and it's just happening because of the underscores.

If someone felt strongly about not changing the behavior (elsewhere), then we could just log a warning instead, when all files are filtered out.

srowen commented:
I think we should implement something along the lines of my comment above: https://github.com/apache/spark/pull/23288#pullrequestreview-185149527

HyukjinKwon commented:
Yup. I think so too.

Commit:
Revert "Clarify the message further with a different exception for file which is ignored"

This reverts commit 08850ae2f64449bae5c449e53c00fa5051479380.

Commit:
Revert "Throws better exception when reading files that start with underscore"

This reverts commit 2910cb9c3c14c97ba4b26d1fab09ab35284aaa3e.

Commit:
Log a debug statement about files/directories that are ignored, and throw an exception only if all of the files are filtered out

KeiichiHirobe commented:
I implemented a new behavior and added tests.
Please check it!

In `checkAndGlobPathIfNecessary`, calling `InMemoryFileIndex.bulkListLeafFiles`.
I am not familiar with cost of calling `InMemoryFileIndex.bulkListLeafFiles`, is this expensive?

If so, maybe we should implement a new method in `object InMemoryFileIndex` which returns information of filtered out files/directories and whether filtered files(regular files) exist or not.

And, please note that I considerd for not only `InMemoryFileIndex.shouldFilterOut` but also `PathFilter.accept`.

HyukjinKwon commented:
Yea, listing files itself is non-trivial in some cases (in particular when you use, for instance, S3) and extra listing up should be avoided. Looks the change is a bit invasive, and does another file listing.

Can we simply do the check when the existing file list happens, rather than doing another file listing?

KeiichiHirobe commented:
Thank you for your reply.

Let me make sure what you mean,
I don't know for certain whether I should list filterOuted files/dirs recursively or  list files/dirs only directly under specified path.
For the example below, when `spark.read.csv("foo")` should we debug log `_a.csv` and `_b.csv` or `_a.csv` only?  

I noticed below behavior for your information.
`bulkListLeafFiles` lists `_a.csv` and `_b.csv`(recursively listed)
`spark.read.csv("foo").show` lists only `a.csv`(read only files directly under specified path)


foo/
　├ a.csv
　├ _a.csv
　├ bar/
　│　├ b.csv
　│　├ _b.csv

srowen commented:
I don't think any of these methods like CSV recurse into subdirectories; you can supply globs that specify files across subdirs. I don't think the path matters here, just the filenames that match. That's what needs to be checked after listing all matching files.

KeiichiHirobe commented:
I get the point.
Looking at https://github.com/apache/spark/pull/23288#discussion_r241663518,
I misunderstood we should check recursively if specified dir.

I'll push later.

Commit:
Change to check only filename match, not check dir recursively

KeiichiHirobe commented:
I pushed now.
Could you please check my commit?

srowen commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -345,6 +346,47 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
   }
 
+  test("SPARK-26339 Debug statement if some of specified paths are filtered out") {
+    class TestAppender extends AppenderSkeleton {
```
I wouldn't bother with this complexity to test if the debug log was printed; it's not important compared to the additional binding to log4j.

KeiichiHirobe replied:
I fixed it.

srowen commented on sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:
```diff
@@ -345,6 +346,47 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils with Te
+    try {
+      val cars = spark
+        .read
+        .format("csv")
```
Nit: you can use `.csv` instead of `format` and `load`

KeiichiHirobe replied:
I fixed it.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:
```diff
@@ -559,6 +559,25 @@ case class DataSource(
     }.toSeq
+
+    if (checkFilesExist) {
+      val (filtered, filteredOut) = allGlobPath.partition { path =>
```
Nit: I'd call `filtered` as `filteredIn` to avoid ambiguity. It might also be very slightly cleaner to avoid the `!` in the expression and flip these two values.

KeiichiHirobe replied:
I fixed it.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:
```diff
@@ -559,6 +559,25 @@ case class DataSource(
+      if (filteredOut.nonEmpty) {
+        if (filtered.isEmpty) {
+          throw new AnalysisException(
+            "All path were ignored. The following path were ignored:\n" +
```
path -> paths. Also, it seems clearer to say: "All paths were ignored:\n" and below, "Some paths were ignored:\n"

KeiichiHirobe replied:
I fixed it.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:
```diff
@@ -559,6 +559,25 @@ case class DataSource(
+              s"${filteredOut.mkString("\n  ")}")
+        } else {
+          logDebug(
+            "The following path were ignored:\n" +
```
Nit: for performance, make this one interpolated string. If the line is too long make the variable `filteredOut` something shorter like `out`

KeiichiHirobe replied:
I fixed it.

srowen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:
```diff
@@ -559,6 +559,25 @@ case class DataSource(
       globPath
     }.toSeq
+
+    if (checkFilesExist) {
```
Do you need to remove the check and exception a few lines above then? It would fail if any path didn't have some files. (Also feel free to fix the indentation from line 549-558 above)

KeiichiHirobe replied:
I am not sure, but irregular indentations seems to be due to GitHub preview CSS.

KeiichiHirobe replied:
No need, I fixed it.

srowen replied:
Yes I mean line 557. I guess we can keep that because, overall, we are trying to throw `AnalysisException` in more cases, not fewer. Before, if one of several glob paths matched no files at all (underscore or not) it would throw. OK, that behavior we can keep, I guess, or at least that's a separate question.

Disregard this; I think it is OK.

Commit:
Avoid checkFilesExist check

Commit:
Minor modifications for clarity

Commit:
Remove debugStatement test

Commit:
Fix title of test case

srowen commented:
Merged to master

HyukjinKwon commented:
@KeiichiHirobe, mind opening a PR again please? I also missed that the test didn't actually run. Looks the current change breaks another regression test. Can you take a look and fix it together?

srowen commented:
Ack, darn, thank you. I was looking at a bunch of open PRs and probably looked at the wrong one to see if tests had run.

KeiichiHirobe commented:
@HyukjinKwon @srowen 
According to [this](https://stackoverflow.com/questions/12674304/github-reopening-a-merged-pull-request), we can not reopen pull request, so I created a new pull request #23446 . Could you please review my commit at #23446?
